spec/rubocop/cop/layout/first_method_parameter_line_break_spec.rb

```diff
@@ -83,4 +83,66 @@ def foo(
       end
     RUBY
   end
+
+  context 'last element can be multiline' do
+    let(:cop_config) { { 'AllowMultilineFinalElement' => true } }
+
+    it 'ignores last argument that value is a multine Hash' do
+      expect_no_offenses(<<~RUBY)
+        def foo(bar, baz = {
+          a: b
+        })
+          do_something
+        end
+      RUBY
+    end
+
+    it 'ignores single argument that value is a multiline hash' do
+      expect_no_offenses(<<~RUBY)
+        def foo(bar = {
+          a: b
+        })
+          do_something
+        end
+      RUBY
+    end
+
+    it 'registers and corrects parameters that value is a multiline hashe and is not the last parameter' do
+      expect_offense(<<~RUBY)
+        def foo(bar, baz = {
+                ^^^ Add a line break before the first parameter of a multi-line method parameter list.
+          a: b
+        }, qux = false)
+          do_something
+        end
+      RUBY
+
+      expect_correction(<<~RUBY)
+        def foo(
+        bar, baz = {
+          a: b
+        }, qux = false)
+          do_something
+        end
+      RUBY
+    end
+
+    it 'registers and corrects last parameter that starts on another line' do
+      expect_offense(<<~RUBY)
+        def foo(bar, baz,
+                ^^^ Add a line break before the first parameter of a multi-line method parameter list.
+          qux = false)
+          do_something
+        end
```

Comment:
I may miss something with this example. It also offends `Layout/ParameterAlignment` which suggests changing to:
```ruby
def foo(bar, baz,
        qux = false)
  do_something
end
```

Does it make sense to use examples that don't trigger other cops?
The primary reason I'm asking is that I'm heavily triggered by this, too, and I feel an irresistible bias against this change in general as a result.

In any case,
```ruby
def foo(
bar, baz,
qux = false)
  do_something
end
```
or the "corrected" one below look horrible to my eye.

Reply:
> or the "corrected" one below look horrible to my eye.

🤔 yeah that rule is definitely not mean to be used in isolation. Not sure what the feedback is here though, this is the one test that is making sure that the existing behaviour is maintained for method definitions that are _actually_ considered multiline even with `AllowMultilineFinalElement` on. It looks a lot like the tests that already existed in this file.

Reply:
@Korri if you run the full correction suite on the original code (with your change) how does it end up looking?

In general, it's better to not have cops do other corrections, especially when it comes to Layout because it results in duplicating a lot of work and additionally causes situations that the correct raises `ClobberingErrors` (the same piece of code is corrected in incompatible ways).

Reply:
> In general, it's better to not have cops do other corrections

What do you mean by "other corrections" the cop here is only adding a new line before the first parameter, it _is_ what it's main and only role is.

For reference, the example you said "looks horrible" differs from what the autocorrect actually does, it doesn't change the indentation of the `qux` keyword, so:
```ruby
# bad
def foo(bar, baz,
        qux = false)
  do_something
end

# get corrected to "good":
def foo(
bar, baz,
  qux = false)
  do_something
end
```

> @Korri if you run the full correction suite on the original code (with your change) how does it end up looking?

Exactly the same as it does today!

First `Layout/FirstMethodParameterLineBreak` is not on by default, and pretty weird to enable on it's own, as it's meant to be used at the very least in conjunction with `Layout/MultilineMethodParameterLineBreaks`, but let's give it a try:

```ruby
# Base example (made a few tweaks to satisfy other rules, but functionally the same)
def foo(bar, baz,
  qux: false)
  do_something(bar, baz, qux)
end

# Autocorrect with the default cops set
def foo(bar, baz,
        qux: false)
  do_something(bar, baz, qux)
end


# With `Layout/FirstMethodParameterLineBreak` enabled
def foo(
  bar, baz,
  qux: false
)
  do_something(bar, baz, qux)
end

# With `Layout/FirstMethodParameterLineBreak` and `Layout/MultilineMethodParameterLineBreaks` enabled:
def foo(
  bar,
  baz,
  qux: false
)
  do_something(bar, baz, qux)
end

# With `Layout/FirstMethodParameterLineBreak` and `Layout/MultilineMethodParameterLineBreaks`  and the `AllowMultilineFinalElement` true on both (no differences):
def foo(
  bar,
  baz,
  qux: false
)
  do_something(bar, baz, qux)
end

```

As the last point shows, there **is NO differences** in behaviour for that specific example, based on the changes in this Pull Request, I only added that unit test to make sure that the existing behaviour DID NOT CHANGE, if you look higher in the test file, you will see _basically the same exact test_.

The current pull request only affects method that have their last element multiline!

Reply:
Thanks for the examples. FYI it wasn't me who you were responding to before.

> What do you mean by "other corrections" the cop here is only adding a new line before the first parameter, it is what it's main and only role is.

Autocorrection for some cops tries to do layout reformatting as well. For instance, adding a new line indented to the correct level. This tends to cause issues in the long run, and is redundant because we have cops that are already present to reflow the layout when code is changed/added.

Reply:
> Thanks for the examples. FYI it wasn't me who you were responding to before.

Hoo totally missed that, will edit to make it clearer!

> Autocorrection for some cops tries to do layout reformatting as well. For instance, a cop that adds a new line might try to indent it to the correct level. This tends to cause issues in the long run, and is redundant because we have cops that are already present to reflow the layout when code is changed/added.

That sounds like exactly what this rule is doing :) It's not trying to fix layout, it's focusing on its one job: Adding a new line.